armasec/token_security.py

```diff
@@ -102,6 +99,22 @@ async def __call__(self, request: Request) -> TokenPayload:
                     detail="Not authenticated",
                     headers={"WWW-Authenticate": "Bearer"},
                 )
+        finally:
```

Comment:
You don't want to put these checks inside of the `finally` clause, because they will be evaluated even after the `except` block above. So, if something handles the exception raised in the `except` block, the user will be very surprised to see that another exception follows closely behind.

Instead, I recommend that you put the checks in the try block above (started line 94) right after the for-loop completes. Then, you can change them to use `require_condition` with `AuthenticationError`:

```python        
try:
    token_payload = self.manager.extract_token_payload(request.headers)
    for domain_config in self.domains_config:
        try:
            manager = self._retrieve_manager(domain_config)
            token_payload = manager.extract_token_payload(request.headers)
            self.manager = manager
            break  # it means there is a matching jwk
        except AuthenticationError:
            self.debug_logger(f"Failed to match JWK against domain {domain_config.domain}")
        except Exception as err:
            if self.debug_exceptions:
                self.debug_logger(f"Exception caught: {err.__class__.__name__}")

    AuthenticationError.require_condition(
        token_payload is not None,
        "Could not find matching JWK with any input domain or token is malformed",
    )
```

This will allow a helpful exception to fall out if you have `debug_exceptions` set.

If more clarification would help here, I would be happy to pair on it with you.

Reply:
>You don't want to put these checks inside of the finally clause, because they will be evaluated even after the except block above. So, if something handles the exception raised in the except block, the user will be very surprised to see that another exception follows closely behind.
The two verifications done in the `finally` step aren't caught by the first try-except block, that's why it's needed in this step because it cannot follow without these checks

Reply:
I also experienced using the require condition the way you suggested, problem is: the API which uses Armasec returns 500, which isn't the best code in this case

Reply:
Nice adjustment. Resolving....